app/models/language.rb

```diff
@@ -51,11 +51,26 @@ def explode_path_to_hash key, path
     {key => path}.merge Hash[(0...elements.size).map{|i|"#{key}[#{i}]".to_sym}.zip(elements)]
   end
 
+  # for validating submissions
   def self.submission_options
-    latest = LanguageGroup.where(identifier: %w[c++ c python haskell java ruby j]).pluck(:current_language_id)
-    old = Language.where(identifier: %w[c++11 c++14 c99 python2]).pluck(:id)
-    languages = Language.where(:id => latest).order(:identifier) + Language.where(:id => old).order(:identifier)
-    Hash[languages.map{ |language| ["#{language.name}", language.id] }]
+    grouped_submission_options.values.reduce(:merge)
+  end
+
+  # for the selection dropdown
+  def self.grouped_submission_options
+    current = ["c++", "c", "python", "java", "haskell", "ruby", "j"]
+    other = ["c++14", "c++11", "c99", "python2"]
+    current_ids = LanguageGroup.where(:identifier => current).pluck(:current_language_id)
+    current_options = Language.where(:id => current_ids).to_a.sort_by { |lang| current.index(lang.group.identifier) }.map { |lang| [lang.name, lang.id] }
```

Comment:
How about `LanguageGroup.where(:identifier => current).map(&:current_language)`? It's less efficient because it loads the entire LanguageGroup model (rather than just the current_language_id column) and makes a separate query for each Language (could be fixed using `.preload(:current_language).map...`), but it's cleaner.

Another approach for preserving the order (not sure which I prefer):
```
current_langs_by_ident = LanguageGroup.where(:identifier => current).map(&:current_language).index_by { |lang| lang.group.identifier }
current_options = current.map { |ident| current_langs_by_ident[ident] }.map { |lang| [lang.name, lang.id] }
```

Reply:
Like this, or is the line getting too long?
```
current_options = LanguageGroup.where(:identifier => current).preload(:current_language).map(&:current_language).to_a.sort_by { |lang| current.index(lang.group.identifier) }.map { |lang| [lang.name, lang.id] }
```

Reply:
I'd say it's too long. I was thinking
```
current_langs = LanguageGroup.where(:identifier => current).map(&:current_language)
current_options = current_langs.sort_by { |lang| current.index(lang.group.identifier) }.map { |lang| [lang.name, lang.id] }
```
`.to_a` isn't necessary after `.map`, and I think it could also be left out in the original version, i.e. `.where(...).sort_by`

Reply:
What does the `&` in `&:current_language` do? It's probably obvious that I don't know ruby. My current strategy is pretty much "hit it until it works and hope to learn something along the way". Thanks for pointing these things out.

Reply:
It [converts a proc to a block](https://ruby-doc.org/core-2.7.0/doc/syntax/calling_methods_rdoc.html#label-Proc+to+Block+Conversion).
`.map` expects a block argument, e.g. `.map { |g| g.current_language }`, so `&` is required to pass the symbol `current_language` as the block argument (rather than a normal positional argument).
Symbols are [convertible to procs](https://ruby-doc.org/core-2.7.0/Proc.html#class-Proc-label-Conversion+of+other+objects+to+procs), and "calling" them on an object calls the object's method with that name.